keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,956 @@
+to track which Kubernetes version the API group-versions are introduced (or
+removed) at.
+
+GA APIs should match the exact set of APIs enabled in GA for the Kubernetes version
+the compatibility version is set to.
+
+All Alpha and Beta APIs (both off-by-default and on-by-default, if any of those
+still exist) need to behave exactly as they did for the Kubernetes version
+the compatibility version is set to. I.e. `--runtime-config=api/<version>` needs
+to be able to turn on APIs exactly like it did for each Kubernetes version that
+compatibility version can be set to.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A cluster administrator is running Kubernetes 1.30.12 and wishes to perform a
+cautious upgrade to 1.31.5 using the smallest upgrade steps possible, validating
+the health of the cluster between each step.
+
+- For each control plane component, in the [recommended
+  order](https://kubernetes.io/releases/version-skew-policy/):
+  - Upgrades binary to `kubernetes-1.31.5` but sets `--compatibility-version=1.30`
+  - Verifies that the cluster is healthy
+- Next, for each control plane component:
+  - Sets `--compatibility-version=1.31`
+  - Verifies that the cluster is healthy
+
+#### Story 2
+
+A cluster administrator is running Kubernetes 1.30.12 and wishes to perform a cautious
+upgrade to 1.31.5, but after upgrading realizes that a feature the workload depends on
+had been removed and needs to rollback until the workload can be modified to not
+depend on the feature.
+
+- For each control plane component, in the [recommended
+  order](https://kubernetes.io/releases/version-skew-policy/):
+  - sets `--compatibility-version=1.30`
+
+This avoids having to rollback the binary version.  Once the workload is fixed, the
+cluster administrator can remove the `--compatibility-version` to roll the cluster
+forward again.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Risk: Introducing this change increases the maintenance burden on Kubernetes
+maintainers.
+
+Why we think this is manageable:
+
+- We already author features to be gated. The only change here is include
+  enough information about features so that they can be selectively enabled/disabled
+  based on compatibility version.
+- We already manually deprecate/remove features. This change will instead
+  leave features in code longer, and require feature gates to track at which
+  verion a feature is deprecated/removed.  The total maintenance work is
```

Comment:
It was mentioned above that the compat version is going to be N-3. With that, is it necessary to mention that tests will be added to make sure the binary version update cannot be more than 3 (may be the processes crash if the difference exceeds).

Reply:
Yes. I'll include this in the integration test plan section.  Our intention is for the process to report a configuration and exit() if the compatibility-version is not in the expected range relative to the binary version.